editor/components/inserter/menu.js

```diff
@@ -12,10 +12,161 @@ import Dashicon from 'components/dashicon';
 class InserterMenu extends wp.element.Component {
 	constructor() {
 		super( ...arguments );
+		this.blockTypes = wp.blocks.getBlocks();
+		this.categories = wp.blocks.getCategories();
+		this.nodes = {};
 		this.state = {
-			filterValue: ''
+			filterValue: '',
+			focusedElementRef: null
 		};
 		this.filter = this.filter.bind( this );
+		this.isShownBlock = this.isShownBlock.bind( this );
+		this.changeMenuSelection = this.changeMenuSelection.bind( this );
+		this.setSearchFocus = this.setSearchFocus.bind( this );
+		this.bindReferenceNode = this.bindReferenceNode.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.listVisibleBlocksByCategory = this.listVisibleBlocksByCategory.bind( this );
+	}
+
+	componentDidMount() {
+		document.addEventListener( 'keydown', this.onKeyDown );
+	}
+
+	componentWillUnmount() {
+		document.removeEventListener( 'keydown', this.onKeyDown );
+	}
+
+	isShown( filterValue, string ) {
+		return string.toLowerCase().indexOf( filterValue.toLowerCase() ) !== -1;
+	}
+
+	isShownBlock( filterValue, block ) {
+		return this.isShown( filterValue, block.title );
+	}
+
+	visibleFocusables( nodes, filterValue ) {
+		const focusables = [];
+
+		for ( const refName in nodes ) {
+			if ( this.isShown( filterValue, refName ) && 'search' !== refName ) {
+				focusables.push( refName );
+			}
+		}
+
+		// Add search regardless at the end.
+		focusables.push( 'search' );
```

Comment:
That looks better to me.  Should we try it with the search input always focused and make it slack style with the /?  Or for now just leave it as is and change everything relative to the blockTypes?